### PR TITLE
Disable "+ NEW" button when cluster is starting

### LINKED ISSUE #103: Visually inform user when we're creating a cluster
When pressing the new cluster button there is currently no feedback to the user that we're doing something.  This is fine if the cluster comes up quickly (as is the case for LocalCluster) but quite stressful when the cluster takes a while (such as is the case when the Scheduler must be run on a remote machine).  In this case the user has the stress of figuring out if they actually pressed the button or not.  If so, they need to wait and not press it again (or else they'll create many clusters).

### PR DESCRIPTION
**Pull request by jsignell**
Closes #103 

After click on "+ NEW", button is disabled and tooltip text changes until cluster is launched:

Screenshot from 2020-02-05 10-48-47

Note that to make this work, I switched from using `ToolbarButton` to `CommandToolbarButton` which allows for easier state updates. This was based on [this](https://gitter.im/jupyterlab/jupyterlab?at=5e399b1c433e1d403995d967) and [this](https://gitter.im/jupyterlab/jupyterlab?at=5e399cf9bfe65274eadcc9c3) comment in gitter.

### CODE REVIEW

File: src/clusters.tsx [hunk 1]
```diff
@@ -45,6 +46,9 @@ export class DaskClusterManager extends Widget {
   /**
    * Create a new Dask cluster manager.
    */
+  status = 'ready';
+  commands = new CommandRegistry()
```
Comment (jsignell) [inline src/clusters.tsx]: Instead of defining a new one, rewrite https://github.com/dask/dask-labextension/blob/master/src/index.ts#L373-L378

Comment (ian-r-rose) [reply src/clusters.tsx]: The command registry is intended to be a application singleton, so we need to either
1. Pass in the command registry as a constructor option to the cluster manager for use in the toolbar (as I think you are suggesting here), or
2. Abandon the idea of a `CommandToolbarButton`, and use the lower-level approach like [here](https://github.com/jupyterlab/jupyterlab/blob/8289dd8b18a04db3889fb69287b617b2677b6338/packages/notebook/src/default-toolbar.tsx#L66-L87).

It's unfortunate that the regular `ToolbarButton` widget can't update its `enabled` state, I think it's a defect in its design. So until we fix that, I think we need one of the two above workarounds.

Comment (ian-r-rose) [reply src/clusters.tsx]: I have already gone through some contortions to avoid the first option (having to pass in the `CommandRegistry`), see, e.g., [here](https://github.com/dask/dask-labextension/blob/4342a1bf088a8b20aff23c30d4e0d52eb6e10641/src/index.ts#L147-L149)

Comment (jsignell) [reply src/clusters.tsx]: Oops I missed this comment. I don't understand why 1 is undesirable.

**Comment by jsignell** [comment]
@ian-r-rose I switched to passing in the commands in the constructor. If you are not opposed to that, then I can switch the rest of the buttons to use the same pattern.

File: src/clusters.tsx (hunk 1 above)
Comment (ian-r-rose) [reply src/clusters.tsx]: Well, it's not exactly a problem to pass in the command registry, so much as an effort to separate concerns. My thinking has been this:

The cluster manager should only need to know about things related to the management of dask clusters. If it has a handle to the command registry, then it can access the whole universe of commands that JupyterLab has, and it becomes more difficult to reason about what the cluster manager does, exactly. That is, if I pass in the command registry, it could call any command that JupyterLab has, sometimes in opaque or confusing ways.

Overall, this is not a huge deal, it's largely me being finicky about APIs.

**Comment by ian-r-rose** [comment]
An aside, can you make sure you have the lint/formatting hooks installed? There are a few minor issues that would be caught by that (e.g., missing semicolons, `!=` vs `!==`).

**Comment by jsignell** [comment]
> An aside, can you make sure you have the lint/formatting hooks installed? There are a few minor issues that would be caught by that (e.g., missing semicolons, `!=` vs `!==`).

I was wondering why I was getting a bunch of missing npm warnings. I hadn't activated my eng. Are there no travis tests for lint?

**Comment by ian-r-rose** [comment]
No, I guess I haven't set those up :)

That would also be a welcome PR, but we can run manually for now.

File: src/clusters.tsx [hunk 2]
```diff
@@ -45,13 +51,19 @@ export class DaskClusterManager extends Widget {
   /**
    * Create a new Dask cluster manager.
    */
+  status = 'ready';
```
Comment (ian-r-rose) [inline src/clusters.tsx]: Can you implement this with a getter and a private attribute? That would make it so `status` couldn't be set by third parties. It would also be nice to tighten up the typings here, maybe something like `DaskClusterManager.IStatus = 'ready' | 'starting' | 'failed'`?

File: src/clusters.tsx [hunk 3]
```diff
@@ -45,13 +51,19 @@ export class DaskClusterManager extends Widget {
...
+  registry: CommandRegistry;
+  launchClusterId: string;
```
Comment (ian-r-rose) [inline src/clusters.tsx]: I think these can be private, and moved down with the rest of the private attributes.

File: src/clusters.tsx (hunk 1 above)
Comment (jsignell) [reply src/clusters.tsx]: No that makes sense. It'd be nice to pass just the one command down to the cluster manager if there needs to be a top level registry. Would you prefer that I use the more react-y approach?

File: src/clusters.tsx [hunk 4]
```diff
@@ -435,10 +445,14 @@ export class DaskClusterManager extends Widget {
     if (response.status !== 200) {
       const err = await response.json();
       void showErrorMessage('Cluster Start Error', err);
+      this.status = 'failed';
```
Comment (ian-r-rose) [inline src/clusters.tsx]: Is there any way to set the status away from 'failed'? I'm not really understanding the semantics: does `ready` mean ready to start a cluster, or that a cluster is ready?

File: src/index.ts [hunk 5]
```diff
@@ -371,9 +373,15 @@ function activate(
 
   // Add a command to launch a new cluster.
   app.commands.addCommand(CommandIDs.launchCluster, {
-    label: 'Launch New Cluster',
-    execute: () => {
-      return sidebar.clusterManager.start();
+    label: 'NEW',
+    execute: () => sidebar.clusterManager.start(),
+    iconClass: 'jp-AddIcon jp-Icon jp-Icon-16',
+    isEnabled: () => sidebar.clusterManager.status != 'starting',
```
Comment (ian-r-rose) [inline src/index.ts]: This should be `!==`. Not sure why the formatter/linter isn't flagging.

File: src/index.ts [hunk 6]
```diff
@@ -371,9 +373,15 @@ function activate(
...
+    caption: () => {
+      if (sidebar.clusterManager.status == 'starting') {
```
Comment (ian-r-rose) [inline src/index.ts]: And here

File: src/clusters.tsx (hunk 1 above)
Comment (ian-r-rose) [reply src/clusters.tsx]: I have a slight preference for the react-y approach, but if it winds up being too much of a pain, I am happy to take the `CommandRegistry` approach.

File: src/clusters.tsx (hunk 4 above)
Comment (jsignell) [reply src/clusters.tsx]: Hmm. I guess it should be just two statuses, `ready` to start a cluster and `starting` all of my checks are just for whether status is starting. So I guess I could also make it a bool. Maybe that would be clearner. `starting: bool`

File: src/clusters.tsx (hunk 3 above)
Comment (jsignell) [reply src/clusters.tsx]: Ok sounds good.

File: src/clusters.tsx (hunk 1 above)
Comment (jsignell) [reply src/clusters.tsx]: I guess since the command exists in the `CommandRegistry` we should use it in the button. So I think I'll leave it as the command-based approach unless you dislike it too much.

File: src/clusters.tsx (hunk 4 above)
Comment (jsignell) [reply src/clusters.tsx]: I went with `isReady` so that the bool matches` isEnabled`.

**Comment by jsignell** [comment]
Should I try the react-y approach, or are you good with this @ian-r-rose?

**Comment by ian-r-rose** [comment]
I think I'm okay with this. I haven't had the chance to take it out for a spin yet. I'll try to do that this evening.

**Comment by jsignell** [comment]
Ok, no rush! Just wanted to let you know that I am done working on it unless you want changes :)

File: src/clusters.tsx [hunk 7]
```diff
@@ -111,16 +120,12 @@ export class DaskClusterManager extends Widget {
...
     toolbar.addItem(
-      'new',
-      new ToolbarButton({
-        iconClassName: 'jp-AddIcon jp-Icon jp-Icon-16',
-        label: 'NEW',
-        onClick: () => {
-          this._launchCluster();
-        },
-        tooltip: 'Start New Dask Cluster'
+      this.launchClusterId,
```
Comment (ian-r-rose) [inline src/clusters.tsx]: This variable name needs updating.

File: src/clusters.tsx [hunk 8]
```diff
@@ -111,16 +120,12 @@ export class DaskClusterManager extends Widget {
...
+      new CommandToolbarButton({
+        commands: this.registry,
+        id: this.launchClusterId
```
Comment (ian-r-rose) [inline src/clusters.tsx]: And here.

File: src/clusters.tsx [hunk 9]
```diff
@@ -111,16 +120,12 @@ export class DaskClusterManager extends Widget {
...
-      new ToolbarButton({
-        iconClassName: 'jp-AddIcon jp-Icon jp-Icon-16',
-        label: 'NEW',
-        onClick: () => {
-          this._launchCluster();
-        },
-        tooltip: 'Start New Dask Cluster'
+      this.launchClusterId,
+      new CommandToolbarButton({
+        commands: this.registry,
```
Comment (ian-r-rose) [inline src/clusters.tsx]: And here

File: src/index.ts [hunk 10]
```diff
@@ -371,9 +373,15 @@ function activate(
...
   app.commands.addCommand(CommandIDs.launchCluster, {
-    label: 'Launch New Cluster',
-    execute: () => {
-      return sidebar.clusterManager.start();
+    label: 'NEW',
+    execute: () => sidebar.clusterManager.start(),
+    iconClass: 'jp-AddIcon jp-Icon jp-Icon-16',
+    isEnabled: () => sidebar.clusterManager.isReady,
+    caption: () => {
+      if (sidebar.clusterManager.isReady) {
```
Comment (ian-r-rose) [inline src/index.ts]: This condition is reversed

**Comment by jsignell** [comment]
@ian-r-rose I'm sorry I asked you to review it in such a sloppy state! I have no idea how I didn't notice that it didn't even build.

File: src/clusters.tsx (hunk 7 above)
Comment (jsignell) [reply src/clusters.tsx]: Yikes.

**Comment by ian-r-rose** [comment]
Looks good to me, thanks @jsignell!

**Comment by ian-r-rose** [comment]
@jacobtomlinson I don't have a lot of experience using versioneer. Can you give me the 20-second version of how to publish patch releases with it?

**Comment by jacobtomlinson** [comment]
Sure!

**TL;DR**

```
jlpm config set version-tag-prefix ""
jlpm version --patch
git push upstream master && git push upstream master --tags
```

I documented the release procedure [here](https://github.com/dask/dask-labextension#publishing).

**Longer explaination**

Versioneer inspects the git tags and log to calculate the current version number. Then when you run `python setup.py sdist bdist_wheel` or similar it replaces the contents of `<package>/_version.py` with a hard coded version instead because you won't be packaging the `.git` directory.

So generically to make a release you run `git tag <version>` to tag it and then build and publish your package.

Because this project has both a JS and Python component I wanted to ensure both the JS and Python components stay in sync. To do that we use `jlpm version` to increment the version because this will both update the `package.json` and also create the git tag. Then you push the tags up to GitHub. 

Travis will then do the `sdist`/`twine`/`npm` uploading bit for you when it sees a new tag.
